src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+uint32_t
+ep_rt_mono_log_single_type (intptr_t type_id)
+{
+    // If there's no room for another type, flush what we've got
+	if (m_n_bulk_type_value_count == k_max_count_type_values)
+		ep_rt_mono_fire_bulk_type_event();
+
+	EP_ASSERT (m_n_bulk_type_value_count < k_max_count_type_values);
+
+	BulkTypeValue *p_val = &m_rg_bulk_type_values[m_n_bulk_type_value_count];
+
+	MonoType *mono_type = (MonoType*)type_id;
+	MonoClass *klass = mono_class_from_mono_type_internal (mono_type);
+
+    // Clear out p_val before filling it out (array elements can get reused if there
+    // are enough types that we need to flush to multiple events).
+	memset(p_val->rg_type_parameters, 0, 100);
+	if (p_val->s_name)
+		p_val->s_name[0] = '\0';
+	p_val->c_type_parameters = 0;
+
+	// Initialize p_val fixed_sized_data
+	p_val->fixed_sized_data.type_id = (uint64_t)type_id;
+	p_val->fixed_sized_data.module_id = (uint64_t)m_class_get_image (klass);
+	p_val->fixed_sized_data.type_name_id = (0x00FFFFFF & m_class_get_type_token (klass)) | 0x02000000;  // dotnet-pgo ResolveMethodID expects type name ids with a 0x02 mask
+	p_val->fixed_sized_data.flags = 0;
+	p_val->fixed_sized_data.cor_element_type = (uint8_t)mono_type->type;
+
+	// Sets p_val variable sized parameter type data c_type_parameters and rg_type_parameters
+	// associated with arrays or generics and add unique identifiers to rg_type_parameters array
+	// to be recursively batched in the same ep_rt_mono_log_type_and_parameters call
+	if ((mono_type->type == MONO_TYPE_ARRAY) || (mono_type->type == MONO_TYPE_SZARRAY))
+	{
+		MonoArrayType *mono_array_type = mono_type_get_array_type (mono_type);
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_ARRAY;
+		if (mono_type->type == MONO_TYPE_ARRAY)
+		{
+			// Only ranks less than kEtwTypeFlagsArrayRankMax are supported.
+			// Fortunately kEtwTypeFlagsArrayRankMax should be greater than the
+			// number of ranks the type loader will support
+			uint32_t rank = mono_array_type->rank;
+			if (rank < (0x3700 >> 8))
+			{
+				rank <<= 8;
+				p_val->fixed_sized_data.flags |= rank;
+			}
+		}
+
+		// Add array's element's class' unique type identifier to rg_type_parameters array
+		if (m_class_is_byreflike (mono_array_type->eklass))
+			p_val->rg_type_parameters[p_val->c_type_parameters++] = m_class_get_this_arg (mono_array_type->eklass);
+		else
+			p_val->rg_type_parameters[p_val->c_type_parameters++] = m_class_get_byval_arg (mono_array_type->eklass);
+	}
+	else if (mono_type->type == MONO_TYPE_GENERICINST)
+	{
+		MonoGenericInst *class_inst = mono_type->data.generic_class->context.class_inst;
+		p_val->c_type_parameters = class_inst->type_argc;
+		for (int i = 0; i < class_inst->type_argc; i++)
+		{
+			// Add generic inst parameter's unique type identifier to rg_type_parameters array
+			MonoClass *mono_class_inst_type_class = mono_class_from_mono_type_internal (class_inst->type_argv[i]);
+			if (m_class_is_byreflike (mono_class_inst_type_class))
```

Comment:
again, this should be `class_inst->type_argv[i]->byref` not `m_class_is_byreflike(...)`

Reply:
Done, using the helper function.